File: fsharp/tiny-maze/test/solver-test.fs
module solver_test

open Expecto
open tiny_maze.solver

[<Tests>]
let tests =
    testList "solver" [
        testCase "can find way to exit with 3x3 maze" <| fun _ ->
            let maze =
                [ [    S; V(0); V(1) ]
                  [ V(1); V(0); V(1) ]
                  [ V(1); V(0);    E ] ]
            let sol =
                [ [    X; X; V(1) ]
                  [ V(1); X; V(1) ]
                  [ V(1); X;    X ] ]
            Expect.equal
                (solve_maze maze) sol ""

        testCase "can find way to exit with 4x4 maze" <| fun _ ->
            let maze =
                [ [    S; V(0); V(0); V(1) ]
                  [ V(1); V(1); V(0); V(0) ]
                  [ V(1); V(0); V(0); V(1) ]
                  [ V(1); V(1); V(0);    E ] ]
            let sol =
                [ [    X;    X; X; V(1) ]
                  [ V(1); V(1); X; V(0) ]
                  [ V(1); V(0); X; V(1) ]
                  [ V(1); V(1); X;    X ] ]
            Expect.equal
                (solve_maze maze) sol ""
    ]

[<EntryPoint>]
let main argv =
    runTestsInAssembly defaultConfig argv

File: csharp/tiny-maze/test/solver-test.cs
namespace tiny_maze;

using Xunit;
using System.Collections.Generic;
using static tiny_maze.solver;

public class solver_test {
    [Fact]
    public void can_exit_3x3_maze() {
        var maze =
            new List<List<P>>() {
                new List<P>() { P.S, P.O, P.W },
                new List<P>() { P.W, P.O, P.W },
                new List<P>() { P.W, P.O, P.E }
            };
        var sol =
            new List<List<P>>() {
                new List<P>() { P.X, P.X, P.W },
                new List<P>() { P.W, P.X, P.W },
                new List<P>() { P.W, P.X, P.X }
            };
        Assert.Equal (sol, solve_maze(maze));
    }

    [Fact]
    public void can_exit_4x4_maze() {
        var maze =
            new List<List<P>>() {
                new List<P>() { P.S, P.O, P.O, P.W },
                new List<P>() { P.W, P.W, P.O, P.O },
                new List<P>() { P.W, P.O, P.O, P.W },
                new List<P>() { P.W, P.W, P.O, P.E },
            };
        var sol =
            new List<List<P>>() {
                new List<P>() { P.X, P.X, P.X, P.W },
                new List<P>() { P.W, P.X, P.X, P.O },
                new List<P>() { P.W, P.O, P.X, P.W },
                new List<P>() { P.W, P.W, P.X, P.X },
            };
        Assert.Equal (sol, solve_maze(maze));
    }
}
